drivers/eeprom/eeprom_fm24w256.c

```diff
@@ -0,0 +1,192 @@
+
+// Get the lower eight address bits and mask them properly. Redundacy for consistency.
+#define OFFSET_LO(offset) ((offset >> 0) & BIT_MASK(8))
+
+LOG_MODULE_REGISTER(fm24w256, CONFIG_EEPROM_LOG_LEVEL);
+
+struct fm24w256_config {
+	struct i2c_dt_spec i2c_spec;
+	struct gpio_dt_spec wp_spec;
+	bool read_only;
+};
+
+struct fm24w256_data {
+	struct k_mutex lock;
+};
+
+static inline const struct fm24w256_config *get_config(const struct device *dev)
+{
+	return dev->config;
+}
+
+static inline struct fm24w256_data *get_data(const struct device *dev)
+{
+	return dev->data;
+}
+
+static inline int enable_write_protect(const struct device *dev)
+{
+	if (!get_config(dev)->wp_spec.port) {
+		return 0;
+	}
+
+	return gpio_pin_set_dt(&get_config(dev)->wp_spec, 1);
+}
+
+static inline int disable_write_protect(const struct device *dev)
+{
+	if (!get_config(dev)->wp_spec.port) {
+		return 0;
+	}
+
+	return gpio_pin_set_dt(&get_config(dev)->wp_spec, 0);
+}
+
+static int fm24w256_memory_operation(const struct device *dev, off_t offset, const void *buf,
+				     size_t buf_len, bool is_write)
+{
+	int ret;
+
+	uint8_t addr_buf[] = {
+		OFFSET_HI(offset),
+		OFFSET_LO(offset),
+	};
+
+	struct i2c_msg msg[] = {
+		{
+			.buf = addr_buf,
+			.len = ARRAY_SIZE(addr_buf),
+			.flags = I2C_MSG_WRITE,
+		},
+		{
+			.buf = (uint8_t *)buf,
+			.len = buf_len,
+			.flags = (is_write ? I2C_MSG_WRITE : I2C_MSG_RESTART | I2C_MSG_READ) |
+				 I2C_MSG_STOP,
+		},
+	};
+
+	if (is_write && get_config(dev)->read_only) {
+		LOG_ERR("Device is configured read only");
+		return -EACCES;
+	}
+
+	if (offset < 0 || offset > SIZE) {
+		LOG_ERR("Offset out of range: %d", offset);
+		return -EINVAL;
+	}
+
+	if (offset + buf_len > SIZE) {
+		LOG_ERR("Memory roll over");
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&get_data(dev)->lock, K_FOREVER);
+
+	ret = disable_write_protect(dev);
+	if (ret) {
+		LOG_ERR("Unable to disable write protection: %d", ret);
+		k_mutex_unlock(&get_data(dev)->lock);
+		return ret;
+	}
+
+	ret = i2c_transfer_dt(&get_config(dev)->i2c_spec, msg, ARRAY_SIZE(msg));
+	if (ret) {
+		LOG_ERR("I2C transfer failed: %d", ret);
+	}
+
+	ret = enable_write_protect(dev);
+	if (ret) {
+		LOG_ERR("Unable to enable write protection: %d", ret);
+	}
+
+	k_mutex_unlock(&get_data(dev)->lock);
+
+	return ret;
+}
+
+static int fm24w256_init(const struct device *dev)
+{
+	int ret;
+
+	ret = k_mutex_init(&get_data(dev)->lock);
+	if (ret) {
+		LOG_ERR("Unable initialize mutex: %d", ret);
+		return ret;
+	}
+
+	/*
+	 * Write protect via GPIO is optional. Only configure if set.
+	 */
+	if (get_config(dev)->wp_spec.port) {
+		if (!device_is_ready(get_config(dev)->wp_spec.port)) {
+			LOG_ERR("write protect port not ready");
+			return -EINVAL;
+		}
+
+		ret = gpio_pin_configure_dt(&get_config(dev)->wp_spec, GPIO_OUTPUT_INACTIVE);
+		if (ret) {
+			LOG_ERR("Unable to configure write protect pin: %d", ret);
+			return ret;
+		}
+	}
+
+	enable_write_protect(dev);
```

Comment:
You could do this in one go by using `GPIO_OUTPUT_ACTIVE`.

Reply:
Done